src/dc_tools.rs

```diff
@@ -415,38 +416,116 @@ pub(crate) fn dc_create_incoming_rfc724_mid(
     match grpid {
-        Some(grpid) => format!("Gr.{}.{}{}", grpid, dc_create_id(), hostname),
-        None => format!("Mr.{}.{}{}", dc_create_id(), dc_create_id(), hostname),
+        Some(grpid) => format!(
+            "{}{}.{}{}",
+            if use_rfc724_msgid_prefix {
+                "chat$group."
+            } else {
+                "Gr."
+            },
+            grpid,
+            dc_create_id(),
+            hostname
+        ),
+        None => format!(
+            "{}{}.{}{}",
+            if use_rfc724_msgid_prefix {
+                "chat$"
+            } else {
+                "Mr."
+            },
+            dc_create_id(),
+            dc_create_id(),
+            hostname
+        ),
     }
 }
 
+/// Strip off `prefix` from `s` and return the remaining string if `s` starts with `prefix`.  If
+/// string `s` does not start with `prefix`, return None.
+///
+/// # Examples
+/// ```rust
+/// use deltachat::dc_tools::split_prefix;
+///
+/// let grp_id = "Gr.12345678901.morerandom@domain.de";
+/// assert_eq!(split_prefix(grp_id, "Gr."), Some("12345678901.morerandom@domain.de"));
+/// assert_eq!(split_prefix(grp_id, "chat$group."), None);
+/// ```
+pub fn split_prefix<'a>(s: &'a str, prefix: &str) -> Option<&'a str> {
+    if s.starts_with(prefix) {
+        s.get(prefix.len()..)
+    } else {
+        None
+    }
+}
+
+#[test]
+fn test_split_prefix() {
+    assert_eq!(split_prefix("", "Gr."), None);
+    assert_eq!(split_prefix("Gr", "Gr."), None);
+    assert_eq!(split_prefix("_Gr.", "Gr."), None);
+    assert_eq!(split_prefix("Gr.", "Gr."), Some(""));
+    assert_eq!(split_prefix("Gr.following", "Gr."), Some("following"));
+    assert_eq!(split_prefix("Gr.Gr.", "Gr."), Some("Gr."));
+}
+
+const OLD_GROUP_ID_PREFIX: &str = "Gr.";
+const NEW_GROUP_ID_PREFIX: &str = "chat$group.";
+
 /// Extract the group id (grpid) from a message id (mid)
 ///
 /// # Arguments
 ///
 /// * `mid` - A string that holds the message id
+///
+/// # Examples
+///
+/// ```rust,ignore
+/// use crate::dc_tools::dc_extract_grpid_from_rfc724_mid;
+/// let mid = "Gr.12345678901.morerandom@domain.de";
+/// let grpid = dc_extract_grpid_from_rfc724_mid(mid);
+/// assert_eq!(grpid, Some("12345678901"));
+/// ```
 pub(crate) fn dc_extract_grpid_from_rfc724_mid(mid: &str) -> Option<&str> {
-    if mid.len() < 9 || !mid.starts_with("Gr.") {
+    if mid.len() < 9 {
         return None;
     }
 
-    if let Some(mid_without_offset) = mid.get(3..) {
-        if let Some(grpid_len) = mid_without_offset.find('.') {
-            /* strict length comparison, the 'Gr.' magic is weak enough */
-            if grpid_len == 11 || grpid_len == 16 {
-                return Some(mid_without_offset.get(0..grpid_len).unwrap());
-            }
-        }
+    if let Some(suffix) = split_prefix(mid, NEW_GROUP_ID_PREFIX) {
+        return suffix
+            .split('.')
+            .next()
+            .filter(|new_grpid| new_grpid.len() >= 12);
```

Comment:
why can new_grpid.len() nore more restricted? 
see also the doc comment on the other case -- and please leave the documentation line ine that you removed.

Reply:
@hpk42 : I don't understand your comment on new_grpid.len(). There were no comments on the last pull request. Do you mean the "strict length comparison, the 'Gr.' magic is weak enough"?

Reply:
yes, i meant that one comment about strict length comparison. i didn't dive deeply but it sounds like there were length checks and in the chat$ they are not done, am i seeing it wrong?